### ISSUE TITLE
The heat prediction function failed, and the prediction result remained 0.

### ISSUE THREAD

**Issue by wjhgq** [666ghj/Weibo_PublicOpinion_AnalysisSystem#6]
Hello author, after I deployed the project, I found that when I used the prediction function, the prediction result kept being 0 or a very large value, which seemed to have some problems. After reviewing the code, I found that you used a logistic regression method and only used 20 recent data points. Does such a strategy lead to a large error? I think there is room for improvement. I have submitted pr to improve this prediction function and hope to pass it.☺️

**Comment by 666ghj** [comment]
Thank you very much for your PR and your valuable suggestions! We’ve indeed been exploring more sophisticated time series models to improve the accuracy of our prediction function. Your contribution is timely and appreciated. I will review your PR this afternoon and provide feedback later today. Thank you for your patience and support!

**Comment by 666ghj** [comment]
A new version has been released, the prediction model has been replaced with a time series model, you can update and try it out.

**Comment by arnvjshi** [comment]
# 🛠️ Issue: Prediction Function Returning 0 or Large Values  

## 🔍 Problem  
After deployment, the prediction function consistently returns either **0 or an unusually large value**. Upon reviewing the code, it was found that:  
- The function **uses only the 20 most recent data points**, which may not capture long-term trends.  
- It **relies on logistic regression**, which is primarily for classification and may not be the best choice for numerical predictions.  

## ✅ Proposed Solution  
### 1️⃣ Expand Data Window Dynamically  
Instead of a fixed 20-point window, we dynamically adjust based on data variance:  

```python
import numpy as np

def get_dynamic_window(data, min_size=20, max_size=100):
    std_dev = np.std(data)
    return min(max_size, max(min_size, int(std_dev * 5)))  # Adaptive window size
```

### 2️⃣ Replace Logistic Regression with Linear Regression  
For more accurate numerical predictions:  

```python
from sklearn.linear_model import LinearRegression

def predict_next_value(data):
    window_size = get_dynamic_window(data)
    X = np.arange(len(data[-window_size:])).reshape(-1, 1)
    y = np.array(data[-window_size:])

    model = LinearRegression()
    model.fit(X, y)
    
    next_value = model.predict([[len(data)]])[0]
    return max(0, next_value)  # Ensure no negative predictions
```

### 3️⃣ Consider Using an LSTM for Time-Series Predictions  
If the dataset has complex trends, an **LSTM model** may significantly improve accuracy.  

Would you like an optimized LSTM implementation for better performance? 🚀  

---

**Comment by 666ghj** [comment]
> # 🛠️ Issue: Prediction Function Returning 0 or Large Values
> ## 🔍 Problem
> After deployment, the prediction function consistently returns either **0 or an unusually large value**. Upon reviewing the code, it was found that:
> 
> * The function **uses only the 20 most recent data points**, which may not capture long-term trends.
> * It **relies on logistic regression**, which is primarily for classification and may not be the best choice for numerical predictions.
> 
> ## ✅ Proposed Solution
> ### 1️⃣ Expand Data Window Dynamically
> Instead of a fixed 20-point window, we dynamically adjust based on data variance:
> 
> import numpy as np
> 
> def get_dynamic_window(data, min_size=20, max_size=100):
>     std_dev = np.std(data)
>     return min(max_size, max(min_size, int(std_dev * 5)))  # Adaptive window size
> ### 2️⃣ Replace Logistic Regression with Linear Regression
> For more accurate numerical predictions:
> 
> from sklearn.linear_model import LinearRegression
> 
> def predict_next_value(data):
>     window_size = get_dynamic_window(data)
>     X = np.arange(len(data[-window_size:])).reshape(-1, 1)
>     y = np.array(data[-window_size:])
> 
>     model = LinearRegression()
>     model.fit(X, y)
>     
>     next_value = model.predict([[len(data)]])[0]
>     return max(0, next_value)  # Ensure no negative predictions
> ### 3️⃣ Consider Using an LSTM for Time-Series Predictions
> If the dataset has complex trends, an **LSTM model** may significantly improve accuracy.
> 
> Would you like an optimized LSTM implementation for better performance? 🚀

This issue seems to have been improved already, but you make a valid point. I'll consider further optimizing it in the next version. You're also very welcome to submit a PR!!

**Comment by 666ghj** [comment]
> # 🛠️ Issue: Prediction Function Returning 0 or Large Values
> ## 🔍 Problem
> After deployment, the prediction function consistently returns either **0 or an unusually large value**. Upon reviewing the code, it was found that:
> 
> * The function **uses only the 20 most recent data points**, which may not capture long-term trends.
> * It **relies on logistic regression**, which is primarily for classification and may not be the best choice for numerical predictions.
> 
> ## ✅ Proposed Solution
> ### 1️⃣ Expand Data Window Dynamically
> Instead of a fixed 20-point window, we dynamically adjust based on data variance:
> 
> import numpy as np
> 
> def get_dynamic_window(data, min_size=20, max_size=100):
>     std_dev = np.std(data)
>     return min(max_size, max(min_size, int(std_dev * 5)))  # Adaptive window size
> ### 2️⃣ Replace Logistic Regression with Linear Regression
> For more accurate numerical predictions:
> 
> from sklearn.linear_model import LinearRegression
> 
> def predict_next_value(data):
>     window_size = get_dynamic_window(data)
>     X = np.arange(len(data[-window_size:])).reshape(-1, 1)
>     y = np.array(data[-window_size:])
> 
>     model = LinearRegression()
>     model.fit(X, y)
>     
>     next_value = model.predict([[len(data)]])[0]
>     return max(0, next_value)  # Ensure no negative predictions
> ### 3️⃣ Consider Using an LSTM for Time-Series Predictions
> If the dataset has complex trends, an **LSTM model** may significantly improve accuracy.
> 
> Would you like an optimized LSTM implementation for better performance? 🚀

Hey arnvjshi, thanks for bringing up the LSTM idea! You’re spot on about its potential for handling complex trends in time-series data. Good news—we’ve actually just integrated an optimized LSTM model into the system with the latest update. It’s built with a bidirectional LSTM to capture long-term dependencies better, uses the BERT tokenizer for text processing, and supports batch predictions for efficiency. It’s already hooked into the prediction pipeline—just select “LSTM Model” on the sentiment prediction page to try it out. That said, there’s always room to refine it further. If you’ve got specific optimizations in mind, I’d love to hear them—feel free to jump in with a PR too!
